### PR TITLE
Migrate to Prost

### PR DESCRIPTION
**Pull request by nrc**
~~Based on #41, lets not land this until after #41.~~

@Hoverbear PTAL at commit https://github.com/tikv/client-rust/commit/24e12022eb5ebb30239711e434d3602dc2084fc1

### CODE REVIEW

File: src/rpc/client.rs [hunk 1]
```diff
@@ -487,7 +487,7 @@ impl RegionContext {
 
 impl From<RegionContext> for kvrpcpb::Context {
     fn from(mut ctx: RegionContext) -> kvrpcpb::Context {
-        let mut kvctx = kvrpcpb::Context::new();
+        let mut kvctx = kvrpcpb::Context::new_();
```
Comment (Hoverbear) [inline src/rpc/client.rs]: Oh no these ugly `new_` again :(

File: examples/raw.rs [hunk 2]
```diff
@@ -1,26 +1,18 @@
...
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// See the License for the specific language governing permissions and
-// limitations under the License.
+// Copyright 2018 TiKV Project Authors. Licensed under Apache-2.0.
+
+#![feature(async_await, await_macro)]
```
Comment (Hoverbear) [inline examples/raw.rs]: Do we have an ETA on stable?

File: examples/transaction.rs [hunk 3]
```diff
@@ -1,77 +1,65 @@
...
-    let _: Vec<()> = future::join_all(
-        pairs
-            .into_iter()
-            .map(Into::into)
-            .map(|p| txn.set(p.key().clone(), p.value().clone())),
-    )
-    .wait()
+    await!(future::join_all(pairs.into_iter().map(Into::into).map(
+        |p| txn.set(p.key().clone(), p.value().clone()).into_future()
+    ),))
```
Comment (Hoverbear) [inline examples/transaction.rs]: The `,` here is really weird.

File: src/compat.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (Hoverbear) [inline src/compat.rs]: Are these kinks that will get worked out before stability?

File: examples/transaction.rs [hunk 5]
```diff
@@ -1,77 +1,65 @@
...
     txn.set_isolation_level(IsolationLevel::ReadCommitted);
-    let _: Vec<()> = keys
-        .into_iter()
-        .map(|p| {
-            txn.delete(p).wait().expect("Could not delete key");
-        })
-        .collect();
-    txn.commit().wait().expect("Could not commit transaction");
+    let _: Vec<()> = await!(stream::iter(keys.into_iter())
+        .then(|p| txn.delete(p).unwrap_or_else(|_| panic!()))
```
Comment (Hoverbear) [inline examples/transaction.rs]: Please include an  error message

**Comment by Hoverbear** [comment]
@nrc Are you proposing to merge this now, or when stable?

File: examples/raw.rs (hunk 2 above)
Comment (nrc) [reply examples/raw.rs]: August 15th. Note that I only use unstable features in the examples, the library itself is all stable (well, technically it needs beta, but will all be stable at the next release).

**Comment by nrc** [comment]
> Are you proposing to merge this now, or when stable?

Now, but I think this and some of the other comments are better discussed on #41. This PR should only be the Prost migration (24e1202) once it is rebased.

File: src/compat.rs (hunk 4 above)
Comment (nrc) [reply src/compat.rs]: No. These are fundamental differences between futures 0.1 and 0.3. They can be removed once we migrate the upstream (grpc) code to futures 0.3.

**Comment by nrc** [comment]
All changes made on #41. I'll rebase this once #41 lands.

File: examples/raw.rs (hunk 2 above)
Comment (Hoverbear) [reply examples/raw.rs]: Ok. I think it's fine if we make this nightly-only for now as long as we are on the road to stable.

File: src/compat.rs (hunk 4 above)
Comment (Hoverbear) [reply src/compat.rs]: Ok. 👍

**Comment by Hoverbear** [comment]
Sorry @nrc xD

**Comment by nrc** [comment]
Rebased to master (no longer based on #41).

PTAL @Hoverbear @ice1000

**Pull request merged**